MediaStore/app/src/main/java/com/android/samples/mediastore/MainActivityViewModel.kt

```diff
@@ -185,6 +200,48 @@ class MainActivityViewModel(application: Application) : AndroidViewModel(applica
         return images
     }
 
+    private suspend fun performDeleteImage(image: MediaStoreImage) {
+        // Starting a delete clears the previous pending image.
+        pendingDeleteImage = image
+
+        withContext(Dispatchers.IO) {
+            try {
+                /**
+                 * In [Build.VERSION_CODES.Q] and above, it isn't possible to modify
+                 * or delete items in MediaStore directly, and explicit permission
+                 * must usually be obtained to do this.
+                 *
+                 * The way it works is the OS will throw a [RecoverableSecurityException],
+                 * which we can catch here. Inside there's an [IntentSender] which the
+                 * activity can use to prompt the user to grant permission to the item
+                 * so it can be either updated or deleted.
+                 */
+                context.contentResolver.delete(
+                    image.contentUri,
+                    "${MediaStore.Images.Media._ID} = ?",
+                    arrayOf(image.id.toString())
+                )
+
+                // Delete was successful.
+                pendingDeleteImage = null
```

Comment:
Again, a bit racey.

Reply:
Updated a bit so hopefully it's less so.